static/src/javascripts/projects/common/modules/identity/api.ts

```diff
@@ -0,0 +1,431 @@
+import { storage } from '@guardian/libs';
+import { mergeCalls } from 'common/modules/async-call-merger';
+import {
+	createAuthenticationComponentEvent,
+	createAuthenticationComponentEventParams,
+} from 'common/modules/identity/auth-component-event-params';
+import config_ from 'lib/config';
+import { getCookie as getCookieByName } from 'lib/cookies';
+import fetchJson from 'lib/fetch-json';
+import mediator from 'lib/mediator';
+import { getUrlVars } from 'lib/url';
+
+// This is really a hacky workaround ⚠️
+const config = config_ as {
+	get: (s: string) => string;
+};
+
+// Types info coming from https://github.com/guardian/discussion-rendering/blob/fc14c26db73bfec8a04ff7a503ed9f90f1a1a8ad/src/types.ts
+
+type PasswordCredential = {
+	id: string;
+	password: string;
+};
+
+type SettableConsent = {
+	id: string;
+	consented: boolean;
+};
+
+type Newsletter = {
+	id: string;
+	subscribed?: boolean;
+};
+
+type UserNameError = {
+	message: string;
+	description: string;
+	context: string;
+};
+
+type UserConsents = {
+	id: string;
+	actor: string;
+	version: number;
+	consented: boolean;
+	timestamp: string;
+	privacyPolicyVersion: number;
+};
+
+type UserGroups = {
+	path: string;
+	packageCode: string;
+};
+
+export type IdentityUser = {
+	dates: { accountCreatedDate: string };
+	consents: UserConsents[];
+	userGroups: UserGroups[];
+	publicFields: {
+		username: string;
+		displayName: string;
+	};
+	statusFields: {
+		userEmailValidated: boolean;
+	};
+	privateFields: {
+		brazeUuid: string;
+		googleTagId: string;
+		legacyPackages: string;
+		legacyProducts: string;
+	};
+	primaryEmailAddress: string;
+	id: string;
+	hasPassword: boolean;
+};
+
+type IdentityUserFromCache = {
+	dates: { accountCreatedDate: string };
+	publicFields: {
+		displayName: string;
+	};
+	statusFields: {
+		userEmailValidated: boolean;
+	};
+	primaryEmailAddress: string;
+	id: string;
+	rawResponse: string;
+} | null;
+
+type IdentityResponse = {
+	status: 'ok' | 'error';
+	user: IdentityUser;
+	errors?: UserNameError[];
+};
+
+let userFromCookieCache: IdentityUserFromCache = null;
+
+const cookieName = 'GU_U';
+const signOutCookieName = 'GU_SO';
+const fbCheckKey = 'gu.id.nextFbCheck';
+let idApiRoot = '';
+let profileRoot = '';
+
+export const init = (): void => {
+	idApiRoot = config.get('page.idApiUrl');
+	mediator.emit('module:identity:api:loaded');
+	profileRoot = config.get('page.idUrl');
+};
+
+export const decodeBase64 = (str: string): string =>
```

Comment:
It looks like this function is defined and used in another part of the codebase toohttps://github.com/guardian/frontend/search?q=decodeBase64 - is it worth moving it to a util somewhere ?

Reply:
Actually, for Javascript, it’s only<sup>1</sup> in the Identity API–and its archived flow files: https://github.com/guardian/frontend/search?l=JavaScript&q=decodeBase64

<sup>1</sup>: _I feel like https://github.com/guardian/frontend/blob/5b970cd7308175cfc1bcae2d4fb8c06ee13c5fa0/common/app/templates/inlineJS/nonBlocking/getUserData.scala.js is a special case._

Reply:
ah cool, I'm happy to leave it in then : )